File: js/Image.js
function ImageMark(num, ctx) {  return function (done) {
    var i = 0;

    var loop = function () {
        var x = Math.floor(Math.random() * window.innerWidth);
        var y = Math.floor(Math.random() * window.innerHeight);
        var w = Math.floor(Math.random() * window.innerWidth);
        var h = Math.floor(Math.random() * window.innerHeight);
        var theta = Math.random() * 2 * Math.PI;

        var img = new Image();

      	img.onload = function() {
              ctx.beginPath();
              ctx.save();
              ctx.rotate(theta);
              ctx.drawImage(img, x, y, w, h);
              ctx.closePath();
              ctx.restore();

            	i++;
            	if (i > 100) {
            	    done();
            	} else {
            	    loop();
            	}

      	};
	    img.onerror = function(e) { console.log(e); }
        img.src = "../images/cc.gif"
    }
    loop();
}}

File: hs/Image.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE ParallelListComp #-}
module Image (benchmark, summary) where

import           Control.Monad

import qualified Data.Text as T
import           Data.Text (Text)

import           Graphics.Blank

import           System.FilePath
import           System.Random

import           Utils

benchmark :: CanvasBenchmark
benchmark ctx = do
    xs     <- replicateM numImages $ randomXCoord ctx
    ys     <- replicateM numImages $ randomYCoord ctx
    ws     <- replicateM numImages $ randomXCoord ctx
    hs     <- replicateM numImages $ randomYCoord ctx
    thetas <- replicateM numImages $ randomRIO (0, 2*pi)
    send' ctx $ do
       img <- newImage image
       sequence_ [ drawTheImage (x,y,w,h) theta img
                         | x     <- xs
                         | y     <- ys
                         | w     <- ws
                         | h     <- hs
                         | theta <- thetas
                         ]

summary :: String
summary = "ImageMark"

numImages :: Int
numImages = 1000

image :: Text
image = T.pack $ "/images" </> "cc" <.> "gif"

type Angle = Double

drawTheImage :: Path -> Angle -> CanvasImage -> Canvas ()
drawTheImage (x,y,w,h) theta img = do
    beginPath();
    save();
    rotate(theta);
    drawImageSize(img, x - (w/2), y - (w/2), w, h);
    closePath();
    restore();
